cdap-api/src/main/java/io/cdap/cdap/api/app/Application.java

```diff
@@ -32,4 +32,17 @@
    * @param context Used to access the environment, application configuration, and application (deployment) arguments
    */
   void configure(ApplicationConfigurer configurer, ApplicationContext<T> context);
+
+  /**
+   * Returns if application supports config upgrade or not.
+   */
+  boolean allowConfigUpgrade();
```

Comment:
is this needed? I'm wondering whether we can just rely on a default implementation for upgradeConfig()

Reply:
This check would be helpful in case when we want to detect if we can upgrade this application or not, like DataStreams or CDC (for now). If we relying on default implementation (where it is idempotent), we can tell users if this application does allow upgrade or not (did it do something or not?) I answered my thoughts on default implementation below but tl;dr is it would be better for developer and users both to clearly know if this is supported or not IMO.

Reply:
so it's just whether it supports upgrade at all then? Let's rename to isUpgradeAllowed() then, since it's more about blocking upgrade than whether we can upgrade the config.

Reply:
Done.